src/list.h

```diff
@@ -0,0 +1,169 @@
+typedef struct list_head list_t;
+
+#define LIST_HEAD_INIT(name)                     \
+	{                                        \
+		.next = &(name), .prev = &(name) \
+	}
+
+#define LIST_HEAD(name) struct list_head name = LIST_HEAD_INIT(name)
+
+#define INIT_LIST_HEAD(ptr)          \
+	({                           \
+		(ptr)->next = (ptr); \
+		(ptr)->prev = (ptr); \
+	})
+
+static inline int
+list_empty(const struct list_head *head)
+{
+	return head->next == head;
+}
+
+static inline void
+__list_add(struct list_head *new, struct list_head *prev,
+           struct list_head *next)
+{
+	next->prev = new;
+	new->next = next;
+	new->prev = prev;
+	prev->next = new;
+}
+
+static inline void
+list_add(struct list_head *new, struct list_head *head)
+{
+	__list_add(new, head, head->next);
+}
+
+static inline void
+list_add_tail(struct list_head *new, struct list_head *head)
+{
+	__list_add(new, head->prev, head);
+}
+
+static inline void
+__list_del(struct list_head *prev, struct list_head *next)
+{
+	next->prev = prev;
+	prev->next = next;
+}
+
+static inline void
+__list_del_entry(struct list_head *entry)
+{
+	__list_del(entry->prev, entry->next);
+}
+
+static inline void
+list_del(struct list_head *entry)
+{
+	__list_del_entry(entry);
+	entry->next = NULL;
+	entry->prev = NULL;
+}
+
+#define list_entry(ptr, type, member) container_of(ptr, type, member)
+
+#define list_first_entry(ptr, type, member) \
+	list_entry((ptr)->next, type, member)
+
+#define list_last_entry(ptr, type, member) list_entry((ptr)->prev, type, member)
+
+#define list_for_each(pos, head) \
+	for (pos = (head)->next; pos != (head); pos = pos->next)
+
+#define list_for_each_safe(pos, n, head)                       \
+	for (pos = (head)->next, n = pos->next; pos != (head); \
+	     pos = n, n = pos->next)
+
+#define list_for_each_prev(pos, head) \
+	for (pos = (head)->prev; pos != (head); pos = pos->prev)
+
+#define list_for_each_prev_safe(pos, n, head)                  \
+	for (pos = (head)->prev, n = pos->prev; pos != (head); \
+	     pos = n, n = pos->prev)
+
+static inline size_t
+list_size(struct list_head *entry)
+{
+	list_t *pos;
+	size_t i = 0;
+	list_for_each(pos, entry) {
+		i++;
+	}
+	return i;
+}
+
+/*
+ * Sort a list with cmp()
+ * creates a temporary array on the heap
+ */
+static inline int
+list_sort(struct list_head *head,
+	  int (*cmp)(const void *a, const void *b, void *state),
+	  void *state)
+{
+	struct list_head **array = NULL, *pos;
+	unsigned int nmemb = 0;
+
+	/*
+	 * This always allocates in page-sized units to avoid realloc actually
+	 * doing anything most of the time.  On 64-bit machines we'll do an
+	 * allocation for every 512 entries.
+	 */
+	const size_t pgcnt = 4096 / sizeof(void *);
+
+	array = calloc(ALIGN_UP(1, pgcnt), sizeof(void *));
+	if (!array)
+		return -1;
+
+	list_for_each(pos, head) {
+		struct list_head **new_array;
+		size_t newmemb = ALIGN_UP(nmemb + 1, pgcnt);
+
+		new_array = reallocarray(array, newmemb, sizeof(*new_array));
```

Comment:
Is it better to do this than calling list_size() you've defined above and using a single allocation?

Reply:
Well, list_size() won't get you there since it's counting entries, not object sizes. We could traverse the entire array twice and make it one allocation, but it's probably not worthwhile: all the use cases for any of the current code I've got wind up being less than 10 pages total.